standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/ReplChangeManager.java

```diff
@@ -205,7 +205,7 @@ private ReplChangeManager(Configuration conf) throws MetaException {
         inited = true;
       }
     } catch (IOException e) {
-      throw new MetaException(StringUtils.stringifyException(e));
+      throw new MetaException(e.getMessage());
```

Comment:
Ya, this ideally should just throw something other than a MetaException, but it wasn't easy to fix like the other location.  Log-and-Throw is a bad practice; the code should do one or the other, not both.  However, if you would like, I can add a log along side the throw.  Log-and-throw is probably a better compromise.

The idea in general though is that MetaException is an Exception that gets passed back to the client via the Thrift API.  It should not be part of any of the core Hive code,... Hive should simply log the exception (`error`) at the Thrift service shim and throw MetaException back to the client.

Reply:
Also, I see I did this log-and-throw in a couple of other places.  Good catch.  Thanks.  Will update.

Reply:
I agree that "Log-and-Throw" is problematic but losing the cause (or part of the stack trace) is equally bad (maybe worse).
I was thinking a bit more about this and came up with the following idea:
```java
throw MetaStoreUtils.newMetaException("Failed to create ReplChangeManager", e);
```
Using the above snippet we can avoid the "Log-and-Throw" anti-pattern and also retain the original cause and stack trace without plumbing everything into the exception message. If ever this error reaches the client it will have a meaningful message pointing the problem without the redundant stack trace.

One problem that I see with my proposal (but also with the proposed changes in this PR to deprecate `StringUtils.stringifyException`) is that if the `MetaException` propagates all the way up to the Thrift processor (i.e., nobody catches it)  then we will have no clue about the real cause; there will be no log events and the error reaching the client will only contain the message (no cause neither stacktrace). Despite this problem, I still believe that the pattern of putting the whole stacktrace as a message in another exception is a bad practice and should not be used.

Moreover, I am not in favor of sending complete stack traces from server to client. I know that there were attempts to explicitly do this (e.g., HIVE-3626, HIVE-26345, etc) in order to give more information to the user but I would advise against. Doing this can leak sensitive information and also goes against the usual guidelines of throwing exceptions appropriate to the abstraction (Effective Java, Item 61: Throw exceptions appropriate to the abstraction).

Long story short, I think it is worth pushing this PR forward possibly incorporating the proposal outlined above. WDYT @belugabehr ?

Reply:
I completely agree with everything you have put forward here.  Checkout out. HIVE-25126.

I do not think however it is such a burden to log-and-throw for now,... with the end goal of removing it completely.  I think adding a utility work-around like this is not ideal in an open-source project because it becomes yet another little thing everyone needs to be aware of, whereas log-and-throw is a much more natural (albeit anti-pattern) behavior.

So to summarize my position:

Yes. Should remove all references to MetaException within the core Hive product.  It should only exists as a log-and-throw at the top-level service level.  It should contain a helpful message for the client in order to respond appropriately to the error condition, and the details (i.e., complete stacktrace) should be logged on the server.  This will making debugging much more consistent.  There have been many times I have been stumped because the stack trace prints on the client side, and just a small message (or nothing at all) appears in the server logs.  It's also confusing because I'm never sure at first if the stacktrace presented on the client side occurred within the server or within the client itself.

Yes. Sending internal stack traces to external clients is a security risk

Yes. `StringUtils.stringifyException` is duplicative of what SLF4J offers and should be removed completely.

Yes. I think we should push through this PR and I can do some more of this work by the same parameters discussed here.  Just waiting on your `+1` :)

Reply:
I am happy that we are on the same page :)

The first thing that came to my mind is doing this:
```java
throw new MetaException("Failed to create ReplChangeManager", e)
```
but it is not possible since MetaException class is generated by Thrift and is missing the relevant constructor.

The next best thing was:
```java
throw (MetaException) new MetaException("Failed to create ReplChangeManager").initCause(e);
```
but it is quite verbose so it feels natural to create a util and that's how I think `MetaStoreUtils.newMetaException` was introduced.

Personally, I would prefer to use `MetaStoreUtils.newMetaException` here but I understand that this is a bit subjective so I do not insist. The points you raised make perfect sense so I am leaving the final decision up to you. You have my `+1` on whatever you pick.